compiler/env/mem/OMRSegmentAllocator.cpp

```diff
@@ -0,0 +1,257 @@
+
+#include "env/mem/SegmentAllocator.hpp"
+#include "env/mem/MemorySegment.hpp"
+#include <algorithm>
+
+
+OMR::SegmentAllocator::SegmentAllocator(TR::BackingMemoryAllocator &backingMemoryAllocator,
+                                        size_t allocationBlockSize)
+   : _backingMemoryAllocator(backingMemoryAllocator)
+   , _allocationBlockSize(allocationBlockSize)
+   , _bytesAllocated(0)
+   , _allocatedSegments( SegmentDequeAllocator(backingMemoryAllocator.rawAllocator()) )
+   , _freeSegments( SegmentDequeAllocator(backingMemoryAllocator.rawAllocator()) )
+   , _currentBackingMemoryChunk( TR::ref(_backingMemoryAllocator.allocate(allocationBlockSize)) )
+   {
+   TR_ASSERT(allocationBlockSize && ((allocationBlockSize & (allocationBlockSize - 1)) == 0), "allocationBlockSize must be power of 2");
+   TR_ASSERT((backingMemoryAllocator.minimumAllocationSize() / allocationBlockSize) * allocationBlockSize == backingMemoryAllocator.minimumAllocationSize(), "allocationBlockSize must evenly divide into backingMemoryAllocator's minimum allocation size");
+   MEMLOG(1, "creation: backing allocator %p block %lu\n", &_backingMemoryAllocator, allocationBlockSize);
+   MEMLOG(2, "chunk allocated %p %lu\n", &(_currentBackingMemoryChunk.get()), _currentBackingMemoryChunk.get().size());
+   }
+
+OMR::SegmentAllocator::~SegmentAllocator() throw()
+   {
+   // delete all the TR::MemorySegments for segments in the allocated list
+   while (!_allocatedSegments.empty())
+      {
+      TR::MemorySegment &segment = _allocatedSegments.back().get();
+      _allocatedSegments.pop_back();
+      _bytesAllocated -= segment.size();
+      _backingMemoryAllocator.rawAllocator().deallocate(&segment, sizeof(TR::MemorySegment));
+      }
+
+   // delete all the TR::MemorySegments for segments in the free list
+   while (!_freeSegments.empty())
+      {
+      TR::MemorySegment &segment = _freeSegments.back().get();
+      _freeSegments.pop_back();
+      _backingMemoryAllocator.rawAllocator().deallocate(&segment, sizeof(TR::MemorySegment));
+      }
+
+   // let _backingMemoryAllocator know to deallocate all its segments
+   _backingMemoryAllocator.deallocateSegments();
+
+   TR_ASSERT(_bytesAllocated == 0, "Unexpectedly _bytesAllocated not equal to zero after freeing all segments");
+   }
+
+TR::MemorySegment &
+OMR::SegmentAllocator::allocate(size_t requiredSize)
+   {
+   // round up to a multiple of the default allocation size
+   size_t const roundedSize = ( ( ( requiredSize + (_allocationBlockSize - 1) ) / _allocationBlockSize ) * _allocationBlockSize );
+   MEMLOG(1, "allocate request %lu will allocate at least %lu\n", requiredSize, roundedSize);
+
+   // if we can recycle a free segment, then do that
+   if (!_freeSegments.empty() && roundedSize == _allocationBlockSize)
+      {
+      TR::MemorySegment &recycledSegment = _freeSegments.back();
+      TR_ASSERT(recycledSegment.size() == roundedSize, "unexpected size on free segment");
```

Comment:
I think in the non-debug build, this should result in a `std::bad_alloc` exception, at least when `recycledSegment.size()` is less than `roundedSize`.

Reply:
fair point: it may also justify `TR_ASSERT_FATAL` since SegmentAllocator is clearly doing something it shouldn't...